runtime/unix.js

```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+   *   | S_FIFO                      (** Named pipe *)
+   *   | S_SOCK                      (** Socket *)
+   */
+  var file_kind;
+  if (js_stats.isFile()) {
+    file_kind = 0;
+  }
+  if (js_stats.isDirectory()) {
+    file_kind = 1;
+  }
+  if (js_stats.isCharacterDevice()) {
+    file_kind = 2;
+  }
+  if (js_stats.isBlockDevice()) {
+    file_kind = 3;
+  }
+  if (js_stats.isSymbolicLink()) {
+    file_kind = 4;
+  }
+  if (js_stats.isFIFO()) {
+    file_kind = 5;
+  }
+  if (js_stats.isSocket()) {
+    file_kind = 6;
+  }
+  /* ===Unix.stats===
+   * type stats =
+   *  { st_dev : int;               (** Device number *)
+   *    st_ino : int;               (** Inode number *)
+   *    st_kind : file_kind;        (** Kind of the file *)
+   *    st_perm : file_perm;        (** Access rights *)
+   *    st_nlink : int;             (** Number of links *)
+   *    st_uid : int;               (** User id of the owner *)
+   *    st_gid : int;               (** Group ID of the file's group *)
+   *    st_rdev : int;              (** Device ID (if special file) *)
+   *    st_size : int;              (** Size in bytes *)
+   *    st_atime : float;           (** Last access time *)
+   *    st_mtime : float;           (** Last modification time *)
+   *    st_ctime : float;           (** Last status change time *)
+   *  }
+   */
+  return BLOCK(
+    0,
+    js_stats.dev,
+    js_stats.ino,
+    file_kind,
+    js_stats.mode,
+    js_stats.nlink,
+    js_stats.uid,
+    js_stats.gid,
+    js_stats.rdev,
+    js_stats.size,
+    js_stats.atimeMs,
+    js_stats.mtimeMs,
+    js_stats.ctimeMs
+  );
+}
+
+//Provides: unix_stat
+//Requires: fs_node_supported, caml_stats_from_js
+//Requires: caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_stat(filename) {
+  if (fs_node_supported()) {
+    var fs = require('fs');
+    var js_filename = caml_jsbytes_of_string(filename);
+    try {
+      var js_stats = fs.statSync(js_filename);
+      return caml_stats_from_js(js_stats);
+    } catch (err) {
+      caml_raise_unix_exn_of_nodejs_error(err);
+    }
+  } else {
+    caml_failwith('unix_stat unsupported outside nodejs');
+  }
+}
+
+//Provides: unix_lstat
+//Requires: fs_node_supported, caml_stats_from_js
+//Requires: caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_lstat(filename) {
+  if (fs_node_supported()) {
+    var fs = require('fs');
+    var js_filename = caml_jsbytes_of_string(filename);
+    try {
+      var js_stats = fs.lstatSync(js_filename);
+      return caml_stats_from_js(js_stats);
+    } catch (err) {
+      caml_raise_unix_exn_of_nodejs_error(err);
+    }
+  } else {
+    caml_failwith('unix_lstat unsupported outside nodejs');
+  }
+}
+
+//Provides: unix_mkdir
+//Requires: fs_node_supported, caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_mkdir(dirname, perm) {
```

Comment:
Related to #1109

Reply:
I've implemented Sys.mkdir in #1134, `unix_mkdir` could directly call into caml_sys_mkdir

Reply:
I investigated this some more and I found that `caml_sys_mkdir` can rethrow a `caml_raise_sys_error`, but we can't reconstruct a `Unix.Unix_error` if done the other way around. So I've refactored a bit to raise the Unix_error and then catch and re-raise in the `caml_sys_*` functions.

Reply:
I've separated that refactor into it's own commits